### PR TITLE
Allow 'private, no-store' Cache-Control header

### LINKED ISSUE #40798: Cache-Control "private, no-store" header not being preserved
https://github.com/rails/rails/pull/39461 changed the `no-store` directive for the `Cache-Control` header to be exclusive, i.e. when setting `Cache-Control` to `private, no-store`, this is simplified to just `no-store`. `private` should typically be superfluous there, but it's not always.

For instance, Fastly "does not currently respect no-store or no-cache directives" and says that "if you need to prevent caching by both Fastly and web browsers, we recommend combining the private directive with max-age=0 or no-store".

https://docs.fastly.com/en/guides/configuring-caching#do-not-cache

It's currently not possible (or at least clear) how to override this directive reduction behaviour.

### Steps to reproduce

Set a `Cache-Control` header of `private, no-store`, e.g.

```ruby
response.headers["Cache-Control"] = "private, no-store"
```

### Expected behavior

The response `Cache-Control` header should have a value of "private, no-store".

### Actual behavior

The response `Cache-Control` header has a value of "no-store".

### System configuration

**Rails version**: 6.1.0

**Ruby version**: 2.6.5

### PR DESCRIPTION
**Pull request by ags**
### Summary

https://github.com/rails/rails/pull/39461 changed the `no-store` directive for the `Cache-Control` header to be exclusive, i.e. when setting `Cache-Control` to `private, no-store`, this is simplified to just `no-store`. `private` should typically be superfluous there, but it's not always.

For instance, Fastly "does not currently respect no-store or no-cache directives" and says that "if you need to prevent caching by both Fastly and web browsers, we recommend combining the private directive with max-age=0 or no-store".

https://docs.fastly.com/en/guides/configuring-caching#do-not-cache

Since it's not possible to override this directive reduction behaviour, the changes in #39461 prevent Fastly users from upgrading Rails.

This changes the behaviour to allow setting a 'private, no-store' header when private is specified - similar to how 'public' can be specified when 'no-cache' is, but not as a default.

Fixes https://github.com/rails/rails/issues/40798

### CODE REVIEW

**Comment by pixeltrix** [comment]
@ags can this not be address by setting the `Surrogate-Control` header to `private`? The [docs on the header][1] only mention `max-age` but could someone check before we go tweaking Rails to workaround issues with CDN providers.

[1]: https://docs.fastly.com/en/guides/configuring-caching#applying-different-cache-rules-for-fastly-and-browsers

**Comment by ags** [comment]
@pixeltrix Unfortunately the [W3C spec](https://www.w3.org/TR/edge-arch/) on `Surrogate-Control` does not list `private` as a valid directive, so that seems unlikely. Fastly's [other docs](https://developer.fastly.com/reference/http-headers/Surrogate-Control/) on the head also don't mention any directives other than `max-age` & `stale-while-revalidate`.
I've emailed Fastly support to confirm this.

**Comment by ags** [comment]
Fastly have replied saying that "by default, we only check Cache-Control header for private value".

**Comment by pixeltrix** [comment]
> Fastly have replied saying that "by default, we only check Cache-Control header for private value".

According to the W3C spec `no-store` is a valid option - do they also ignore that there?

**Comment by PikachuEXE** [comment]
@pixeltrix 
Yes Fastly ignores `no-store`
https://docs.fastly.com/en/guides/configuring-caching#do-not-cache
image

**Comment by pixeltrix** [comment]
> Yes Fastly ignores `no-store`

Is that true for `Surrogate-Control` as well? The documentation is unclear and only mentions ignoring it in `Cache-Control`.

**Comment by PikachuEXE** [comment]
There are [multiple doc entries for `Surrogate-Control`](https://docs.fastly.com/search#stq=Surrogate-Control&stp=1)
But I didn't read all of them yet
To be really sure might as well send a message to them referencing this issue

**Comment by pixeltrix** [comment]
Not trying to be awkward here but the original change in #39461 talks about PCI-DSS compliance so if we're going to change it we'll need to investigate that angle before it's merged.

**Comment by PikachuEXE** [comment]
Speaking of that
I read the MDN article about cache control
And found that [it's fine to use `no-store, max-age=0`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#preventing_caching)

**Comment by pixeltrix** [comment]
> Speaking of that
> I read the MDN article about cache control
> And found that [it's fine to use `no-store, max-age=0`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#preventing_caching)

Does that help us? The Fastly docs talk about using `private` with either `no-store` or `max-age` not replacing it.

**Comment by ags** [comment]
I think https://github.com/rails/rails/pull/39461 was about allowing `no-store` be an exclusive value for `Cache-Control`, which it couldn't be before, which is a useful change to make for that use case. It's also fine for `no-store` *not* to be exclusive, but https://github.com/rails/rails/pull/39461 changed rails such that it always is. This change tries to bridge that gap by continuing to allow `no-store` to be exclusive, but also allowing it to be combined with other directives too.

**Comment by ags** [comment]
> Is that true for Surrogate-Control as well? 

Yes, it appears to be. The default generated VCL of a Fastly service will only use `Surrogate-Control` to determine the initial TTL of responses:

```vcl
  if (beresp.http.Expires || beresp.http.Surrogate-Control ~ "max-age" || beresp.http.Cache-Control ~"(s-maxage|max-age)") {
    # keep the ttl here
  } else {
        # apply the default ttl
    set beresp.ttl = 300s;
  }
```

This behaviour (and that of `Cache-Control`) can be changed through custom VCL, but being able to set it from Rails, as it could be before, would be preferable as it could vary depending on app logic.

**Comment by alipman88** [comment]
Looking at e06f7b3, I believe it changed Rails' behavior beyond what was intended: the intention was to stop automatically appending a `private` option to the Cache-Control header when `no-store` is already present. But it effectively prevents combining `no-store` with any other Cache-Control header options.

I think there are a number of valid reasons one might want to combine `no-store` with other options.

For instance, this blog post recommends using both `no-store` and `no-cache` to work around a Safari bug: https://blog.alex-miller.co/rails/2017/01/07/rails-authenticity-token-and-mobile-safari.html#the-solution

@ags, could your solution be broadened to allow `no-store` to be combined with other Cache-Control header options, not just `private`?

Maybe something like this:
```diff
diff --git a/actionpack/lib/action_dispatch/http/cache.rb b/actionpack/lib/action_dispatch/http/cache.rb
index a50f72ae90..28e3b6dca7 100644
--- a/actionpack/lib/action_dispatch/http/cache.rb
+++ b/actionpack/lib/action_dispatch/http/cache.rb
@@ -195,9 +195,7 @@ def merge_and_normalize_cache_control!(cache_control)

           control.merge! cache_control

-          if control[:no_store]
-            self._cache_control = NO_STORE
-          elsif control[:no_cache]
+          if control[:no_cache]
             options = []
             options << PUBLIC if control[:public]
             options << NO_CACHE
@@ -212,7 +210,7 @@ def merge_and_normalize_cache_control!(cache_control)

             options = []
             options << "max-age=#{max_age.to_i}" if max_age
-            options << (control[:public] ? PUBLIC : PRIVATE)
+            options << (control[:public] ? PUBLIC : PRIVATE) unless control[:no_store]
             options << MUST_REVALIDATE if control[:must_revalidate]
             options << "stale-while-revalidate=#{stale_while_revalidate.to_i}" if stale_while_revalidate
             options << "stale-if-error=#{stale_if_error.to_i}" if stale_if_error
```

**Comment by pixeltrix** [comment]
@alipman88 under the principle of 'sharp knives' I'd agree that letting users configure whatever they want would be the way to go and we should special case `:no_cache` either. So I tried removing the two conditions while just leaving the `else` block with `no-cache` and `no-store` added. Looking at what failed and what would need adjusting leads me to think that would risk breaking too many applications. I'm familiar with the use case you link to as I've commented heavily on the issue linked to in the article (#21948) but it doesn't really fix the issue, just reduces the likelihood and that's because of the `no-store` rather than the `no-cache` option. The only real solution is to switch to `reset_session` and either dump the user back to the login page or if your page is unauthenticated the form submission will then work (that's what I've done here alphagov/e-petitions#778 after a security review objected to a 2 week session cookie).

@ags after much reading of caching articles by the various and testing of differences between `no-store` and `private, no-store` (albeit in Safari only), there doesn't appear to be any difference in behaviour between the two - the `no-store` seems to take precedence and loading the page from the browser's history menu will refetch the page from the server. I thought it might make a difference with clicking the back button but Safari keeps a copy for both versions.

File: actionpack/lib/action_dispatch/http/cache.rb
Comment (pixeltrix) [inline actionpack/lib/action_dispatch/http/cache.rb:203]: Since all conditions have `options` now, we can pull out the `options = []` and `self._cache_control = options.join(", ")` to before and after the `if` block.

**Review by pixeltrix — CHANGES_REQUESTED** [review]

**Comment by pixeltrix** [comment]
@ags if you make the tiny change and rebase to remove the conflict I'll merge it.

**Comment by alipman88** [comment]
@pixeltrix - my app saw a spike in CSRF token errors from mobile Safari users after upgrading to Rails 6.1, going from zero to a couple dozen per day. We previously fixed these errors by setting a `no-store, no-cache` Cache-Control header, so my initial suspicion is that our errors are triggered by the absence of `no-cache`.

Screen Shot 2021-03-31 at 4 32 13 PM

I've written a monkey-patch to restore the previous 6.0 behavior, but I won't know for sure whether it's helping until my next deployment goes out on Monday. I'll open a new issue or PR if I can identify some common scenarios where `no-cache` is needed in conjunction with `no-store` (e.g. particular versions of mobile Safari). In the mean time, I like @ags's patch.

**Comment by pixeltrix** [comment]
@alipman88 my tests today were using desktop Safari so it's possible there maybe a difference in behaviour in mobile Safari so it's worth investigating further.

File: actionpack/lib/action_dispatch/http/cache.rb
Comment (ags) [reply actionpack/lib/action_dispatch/http/cache.rb]: Good suggestion, done!

**Comment by ags** [comment]
> @ags if you make the tiny change and rebase to remove the conflict I'll merge it.

Done, thanks!

**Review by pixeltrix — APPROVED** [review]

**Comment by pixeltrix** [comment]
@ags thanks! 👍🏻

**Comment by rodzyn** [comment]
@ags Are there plans to merge this fix to other branches, like `6-1-stable`?

**Comment by ags** [comment]
@rodzyn I'm not sure, sorry. Maybe @pixeltrix would know?

**Comment by pixeltrix** [comment]
I think it should be okay to backport this since you explicitly need to add the private key to the options.

**Comment by pixeltrix** [comment]
@rodzyn @ags backported to 6-1-stable

**Comment by ags** [comment]
Thanks @pixeltrix!

**Comment by mockdeep** [comment]
@alipman88 did you ever discover the cause of the errors? We're looking at upgrading to Rails 6.1, but we've also got `no-cache, no-store`, and want to make sure things are working on mobile safari.

**Comment by alipman88** [comment]
@mockdeep - I no longer believe the errors were connected to our Rails 6.1 upgrade - sorry for not updating the thread.

(In short, back in April I had recently started on a new team that used a bug tracker, Rollbar, I was unfamiliar with. Rollbar's AI for grouping individual error occurrences into issues occasionally makes it look like a release introduced a new issue when in reality it had already been occurring for some time - that's what was happening here.)

**Comment by mockdeep** [comment]
@alipman88 thanks for the follow-up! I guess we'll hold off on that monkey patch then.

**Pull request merged**
